### ISSUE TITLE
iptables rules do NOT block unencrypted L2TP clients from connecting

### ISSUE THREAD

**Issue by ryt51V** [hwdsl2/setup-ipsec-vpn#116]
tl;dr, steps to reproduce:

1. Run the script, in particular allowing the firewall rules to be created as-is.
2. Bring up an L2TP/IPsec client connection.
3. Disconnect the encrypted connection.
4. Bring up an L2TP client connection without IPsec.
5. You will connect sucessfully.

Long version:

I've discovered that my particular L2TP/IPsec client will sometimes "forget" to use IPsec and initiate an unencrypted L2TP connection.  Now that's completely the client's fault and obviously unrelated to this script, to that end I'm already trying to responsibly disclose that to the vendor ;-)

That being said, the server **should** be blocking unencrypted L2TP connections to help prevent this sort of mistake.  Indeed, [it's clear that this is already the intention](https://github.com/hwdsl2/setup-ipsec-vpn/blob/43d11fe35a6f22a0a5bfa8a140e044bdff5b9f49/vpnsetup.sh#L359-L363).

```
  iptables -I INPUT 1 -m conntrack --ctstate INVALID -j DROP
  iptables -I INPUT 2 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
  iptables -I INPUT 3 -p udp -m multiport --dports 500,4500 -j ACCEPT
  iptables -I INPUT 4 -p udp --dport 1701 -m policy --dir in --pol ipsec -j ACCEPT
  iptables -I INPUT 5 -p udp --dport 1701 -j DROP
```

The problem seems to be INPUT 2 - the related/established rule.  If you remove rule 2 then it works as expected - L2TP/IPsec connections work fine, unencrypted L2TP does not, and you can see the packet count for rule 5 go up when you try.  Put rule 2 back in, and you can connect unencrypted L2TP again, and the rule 5 count remains constant.

Presumably what's happening is as follows:
1. You establish an encrypted connection, which is allowed by rule 4.  It's then added to the connection tracking.
2. You take down the encrypted connection.
3. You bring up an unencrypted connection.  This is related/established and so is allowed by rule 2.
that the encrypted connection is added to the connection tracking after matching rule 4.

Indeed, if you reboot the server to wipe out tracked connections, and then begin an unencrypted L2TP client connection, it can't connect (it doesn't satisfy rule 2, so rule 5 drops it).  But if you then bring it up encrypted, then down, then bring it up unencrypted again, it can connect just fine.

I'm still looking into solutions and workarounds.  We don't want to remove rule 2 outright as it's important for firewall performance.  As a quick fix you can simply reorder the rules:

```
  iptables -I INPUT 1 -m conntrack --ctstate INVALID -j DROP
  iptables -I INPUT 2 -p udp -m multiport --dports 500,4500 -j ACCEPT
  iptables -I INPUT 3 -p udp --dport 1701 -m policy --dir in --pol ipsec -j ACCEPT
  iptables -I INPUT 4 -p udp --dport 1701 -j DROP
  iptables -I INPUT 5 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
```

Now rule 5 will drop it before connection tracking is considered.  However theoretically this will reduce general network performance, as many connections will have to go through 5 rules to be accepted rather than 2, so it's not a best practice.

To avoid any doubt, here is the full iptables I am testing with.  It's a subset of rules from the script plus one unrelated.

```
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [33:5112]
-A INPUT -m conntrack --ctstate INVALID -j DROP
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A INPUT -p udp -m multiport --dports 500,4500 -j ACCEPT
-A INPUT -p udp -m udp --dport 1701 -m policy --dir in --pol ipsec -j ACCEPT
-A INPUT -p udp -m udp --dport 1701 -j DROP
-A INPUT -p tcp -m tcp --dport <completely unrelated port> -j ACCEPT
-A INPUT -i eth0 -j DROP
-A FORWARD -j DROP
COMMIT
```

Thanks for the awesome script by the way  👍

**Comment by hwdsl2** [comment]
@ryt51V First, thanks for the detailed report. To be clear for everyone, the IPTables rules added by the script already protects against any unencrypted L2TP connection from being established.

The issue you mentioned seems to be a rare case where the VPN client will first connect using IPsec/L2TP, then disconnect and immediately connect again using unencrypted L2TP:

> 1. You establish an encrypted connection, which is allowed by rule 4. It's then added to the connection tracking.
> 2. You take down the encrypted connection.
> 3. You bring up an unencrypted connection. This is related/established and so is allowed by rule 2.
>     that the encrypted connection is added to the connection tracking after matching rule 4.

In this uncommon scenario, the VPN client is not behaving properly (why connect using IPsec/L2TP first, then unencrypted L2TP immediately after that? Do you have more details on your particular client setup which will cause this?), and I think the problem is not with the VPN server, but instead that the VPN client should be fixed.

**Comment by ryt51V** [comment]
@hwdsl2 Thanks for your response.

That's pretty much correct, but you don't have to re-connect "immediately" after disconnecting to establish the unencrypted session.  It's however long before conntrack removes the connection on the server, which is somewhere between around 3-5 minutes.  (I've tried to find out the default time in Debian but haven't had much luck - it's usually in /proc somewhere.  [I think it's one of these.](https://www.kernel.org/doc/Documentation/networking/nf_conntrack-sysctl.txt))

I completely agree with you that the VPN client is not behaving properly!  I don't want to go into too many details right now as I am trying responsible disclosure - I have reported this to the vendor and am trying to give them sufficient time to fix this.  But vaguely, it's a hardware device with an L2TP/IPsec client built in, and for some reason when you reboot the device, it will attempt to establish the session but "forget" to use IPsec.  (It will use IPsec again if you drop and reestablish the connection manually after this).  Rebooting is a common reason why you might drop and re-establish the connection in a short period of time.

While it's clearly the VPN client's fault for trying the unencrypted connection, in my opinion the server should also be fully blocking L2TP without IPsec to prevent unknowing mistakes like this - much like how many companies make their websites only work in HTTPS.  But obviously it's your script and your decision whether this is a bug or not :) I just wanted to bring it to your attention; I'm more than happy to figure out a fix myself for my personal situation.

**Comment by hwdsl2** [comment]
@ryt51V Thanks for the update. I looked into it further and found a better workaround for your particular VPN client issue. You can use the IPTables "invert matching" [1] to exclude UDP port 1701 from the "RELATED,ESTABLISHED" rule. This has less impact on the performance I think, and should help address the issue you described.

```
-A INPUT -m conntrack --ctstate INVALID -j DROP
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -p tcp -j ACCEPT
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -p udp ! --dport 1701 -j ACCEPT
-A INPUT -p udp -m multiport --dports 500,4500 -j ACCEPT
-A INPUT -p udp --dport 1701 -m policy --dir in --pol ipsec -j ACCEPT
-A INPUT -p udp --dport 1701 -j DROP
```

[1] https://www.cyberciti.biz/faq/iptables-invert-ip-or-protocol-with/

**Comment by hwdsl2** [comment]
@ryt51V Actually there is a better workaround for this issue. We can just drop the UDP 1701 packets that does not match any IPsec policy, before the "RELATED,ESTABLISHED" rule, like this:

```
-A INPUT -p udp --dport 1701 -m policy --dir in --pol none -j DROP
-A INPUT -m conntrack --ctstate INVALID -j DROP
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A INPUT -p udp -m multiport --dports 500,4500 -j ACCEPT
-A INPUT -p udp --dport 1701 -m policy --dir in --pol ipsec -j ACCEPT
-A INPUT -p udp --dport 1701 -j DROP
```

**Comment by ryt51V** [comment]
@hwdsl2 Thank you very much - I have tested that second ruleset and it works perfectly.
